Improvements to tls API documentation, batch 1

sam-github opened the pull request:
##### Checklist

- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

doc,tls

##### Description of change

Improvements to tls API documentation.

sam-github commented:
@nodejs/crypto @nodejs/documentation @silverwind @jasnell

sam-github commented on doc/api/tls.md:
```diff
@@ -918,81 +890,87 @@ added: v0.11.13
 * `options` {Object}
-  * `pfx` {string|Buffer} A string or `Buffer` containing the private key,
-    certificate and CA certs of the server in PFX or PKCS12 format. (Mutually
-    exclusive with the `key`, `cert`, and `ca` options.)
```
My tests and reading of code show it not mutually exclusive with key/cert, btw, and it would be quite strange if it was mutually exclusive with ca. I am still extending test suite to confirm this.

silverwind commented on doc/api/tls.md:
```diff
@@ -918,81 +890,87 @@ added: v0.11.13
+  * `honorCipherOrder` {boolean} Attempt to use the server's cipher suite
+    preferences instead of the client's. When `true`, causes
+    `SSL_OP_CIPHER_SERVER_PREFERENCE` to be set in `secureOptions`, see
+    [OpenSSL Options][] for more information.
```
Defaults to `true`.

sam-github replied:
Did you read the Note below? It doesn't default to  true if you do `new tls.TLSSocket({isServer: true, ...})`.

sam-github replied:
cf. https://github.com/nodejs/node/blob/master/lib/_tls_wrap.js#L917

silverwind replied:
Not sure I follow. What influence does `isServer` have over `honorCipherOrder`? The linked code seems pretty clear that it is `true` unless user-provided.

sam-github replied:
See my comment two above. The linked code is in `tls.createServer()`, that code isn't run when calling `new tls.TLSSocket()` (server or client variant), and of course not when calling `tls.connect()`, not that it would matter in that case, its a server only option.

Clearly you are finding this confusing, though, and the docs shouldn't be confusing. Any suggestions for making it clear that its true by default when using `tls.createServer()`, but not when using `new tls.TLSSocket({isServer: true, ...})`? In other words, the common way of making server sockets had a default value, but the other supported and documented way of making them does not set it?

I  thought I said that pretty explicitly.

silverwind replied:
I'd argue that it should be the default for all servers, at least that was the intention when I made that change. Should be a pretty low-impact change, no?

sam-github replied:
I agree its odd. In the process of documenting the existing behaviour, I found many odd and irregular aspects of the TLS API, but this PR documents TLS "as is". I definitely encourage you to PR a fix to move the setting of the default out of `function Server()` and into `_tls_common` where it can apply to all TLS sockets, and be based on the `.isServer` property. If you don't, I will. I consider completing the docs to be step 1 to cleaning up the API. In the absence of docs, its hard for reviewers to understand why behaviour is inconsistent and needs to be fixed.

sam-github replied:
I have started a PR to make it default for all servers, FYI, @silverwind, but might not PR it until I'm back from node interactive. Btw, look for me if you are down here.

silverwind replied:
👍 to that, I'm not too familiar around tls, so I welcome you doing that change. It was certainly an oversight when we flipped that option.

Oh, and I'm unfortunately not attending NI.

silverwind commented on doc/api/tls.md:
```diff
@@ -918,81 +890,87 @@ added: v0.11.13
+    [Perfect Forward Secrecy][]. Use `openssl dhparam` to create the parameters.
+    The key length must be greater than or equal to 1024 bits, otherwise an
+    error will be thrown. It is strongly recommended to use 2048 bits or larger
+    for stronger security. If omitted or invalid, the parameters are silently
```
`adequate security`?

sam-github replied:
Verbatim copy of existing text: https://github.com/nodejs/node/blob/1f45d7aa41291383ce63cca5f80fd2a2a73d7603/doc/api/tls.md#tlscreateserveroptions-secureconnectionlistener

sam-github commented on doc/api/tls.md:
```diff
@@ -918,81 +890,87 @@ added: v0.11.13
+  * `ca`{string|string[]|Buffer|Buffer[]} Optional CA certificates to trust.
+    Default is the well-known CAs from Mozilla. When connecting to peers that
+    use certificates issued privately, or self-signed, the private root CA or
+    self-signed certificate must be provided to verify the peer.
```
Isn't that how all default values work? The default is used when you don't specify the option. If you specify the option, the default isn't used.

I can spell it out if necessary, but the other `Optional x. Default is y.` don't.

I think the old text used to be confusing, which is why some people thought it could possibly be additive.

silverwind replied:
Right, you have a point. It's probably okay your way.

sam-github commented on doc/api/tls.md:
```diff
@@ -179,11 +180,33 @@ line switch. For instance, the following makes
+[`tls.TLSSocket`]s.
+
+Consult
+<https://www.openssl.org/docs/man1.0.2/apps/ciphers.html#CIPHER-LIST-FORMAT> for
```
ok, maybe In a follow on commit, because this style of link to openssl docs is pervasive here and in crypto.md, I can change them all in one go.

sam-github commented on doc/api/tls.md:
```diff
@@ -297,11 +320,11 @@ The following illustrates resuming a TLS session:
 });
 server.on('resumeSession', (id, cb) => {
-  cb(null, tlsSessionStore[id.toString('hex')] || null);
+  cb(null, tlsSessionStore[id.tostring('hex')] || null);
```
yes. I was staring and staring at this, but I didn't see the case change. Now I do.

sam-github commented:
Thanks, I'll fix the stuff you found.

mscdex commented:
The number of commits seems a bit much, can't some of the common changes (e.g. 'describe' commits) be squashed together?

sam-github commented:
@mscdex of course I can squash before merge, each commit would become a bullet point. squashing is easy, unsquashing is hard, which is why its like this for now. @jasnell suggested a series of small commits would make it easier to review than a mega-rewrite.

sam-github commented:
@nodejs/security, one of you care to review for accuracy?

@Trott I addressed your feedback, do you have time to review again?

@jasnell or @Trott, opinion on merging as series of commits, or @mscdex 's suggestion of squashing them, perhaps into one per `test:`/`doc:`/`tls:` subsystem?

mscdex commented:
@sam-github For review it's fine as-is, but I just wanted to make sure if/when this PR is landed, that the like-commits get squashed together and not landed as-is.

Trott commented on test/parallel/test-tls-passphrase.js:
```diff
@@ -11,11 +11,17 @@ var tls = require('tls');
 
+assert(Buffer.isBuffer(passKey));
+assert(Buffer.isBuffer(cert));
+assert.equal(typeof passKey.toString(), 'string');
```
Nit: `assert.strictEqual()` instead of `assert.equal()` here and anywhere else it shows up

Trott replied:
Nit: `assert.strictEqual()` instead of `assert.equal()` here and anywhere else it shows up

Trott commented on test/parallel/test-tls-passphrase.js:
```diff
@@ -11,11 +11,17 @@ var tls = require('tls');
 var path = require('path');
 
-var key = fs.readFileSync(path.join(common.fixturesDir, 'pass-key.pem'));
+var passKey = fs.readFileSync(path.join(common.fixturesDir, 'pass-key.pem'));
```
Nit: While in here anyway, maybe take an opportunity to var->const some stuff? Up to you.

Trott commented on test/parallel/test-tls-no-cert-required.js:
```diff
@@ -1,4 +1,5 @@
 'use strict';
+var assert = require('assert');
```
Nit: var->const in this file too if you are so inclined (and not if you are not).

sam-github commented on test/parallel/test-tls-passphrase.js:
```diff
@@ -26,22 +32,172 @@ var server = tls.Server({
+  */
+})).unref();
 
 assert.throws(function() {
```
done

sam-github commented:
Squashed and landed in d4f00fe..db50307

sam-github commented:
I manually cherry-picked this all onto v7.x-staging, and they land clean.

Fishrock123 commented:
@sam-github odd, I can take a look at the bot logs, but I'm not sure they go back that far.

MylesBorins commented:
@sam-github this lands cleanly on `v6.x`, just wanted to verify that this should land. If so please feel free to cherry-pick and push upstream.

sam-github commented:
OK. I diffed v4...v6, and v6...master, and I believe these docs apply equally to v4, v6, and v7.

I will cherry pick them now onto v6.x-staging.

sam-github commented:
@MylesBorins On v6.x-staging:

```
5416926 test: var to const in tls-no-cert-required
58615e5 doc: rework tls for accuracy and clarity
ab9dd9d tls: do not refer to secureOptions as flags
d0ee5e0 test: tls key/cert ordering not necessary
aaa57d5 tls: document and test option-less createServer
e594153 tls: fix/annotate connect arg comments
```
